drivers/flash/flash_qspi_renesas_ra.c

```diff
@@ -0,0 +1,501 @@
+	int err = 0;
+	uint8_t buffer[size];
+
+	acquire_device(dev);
+	memset(&buffer[0], 0, sizeof(buffer));
+	buffer[0] = QSPI_CMD_READ_SFDP;
+	buffer[1] = addr;
+	buffer[2] = addr >> 8;
+	buffer[3] = addr >> 16;
+
+	err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &buffer[0], FOUR_BYTE, true);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: direct write failed");
+		err = -EIO;
+		goto out;
+	}
+
+	err = R_QSPI_DirectRead(&qspi_data->qspi_ctrl, &buffer[0], size);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: direct read failed");
+		err = -EIO;
+		goto out;
+	}
+
+	err = get_flash_status(dev);
+	if (err != FSP_SUCCESS) {
+		LOG_ERR("qspi: failed to get status for QSPI operation");
+		err = -EIO;
+		goto out;
+	}
+
+	if (SPI_FLASH_PROTOCOL_QPI == qspi_data->qspi_cfg.spi_protocol) {
+		/* 3 dummy byte */
+		memcpy(data, &buffer[4], size);
+	} else {
+		/* 1 dummy byte */
+		memcpy(data, &buffer[1], size);
+	}
+out:
+	release_device(dev);
+	return err;
+}
+#endif
+
+static bool qspi_flash_ra_valid(off_t area_size, off_t offset, size_t len)
+{
+	if ((offset < 0) || (offset >= area_size) || ((area_size - offset) < len)) {
+		return false;
+	}
+
+	return true;
+}
+
+static int qspi_flash_ra_erase(const struct device *dev, off_t offset, size_t len)
+{
+	struct qspi_flash_ra_data *qspi_data = dev->data;
+	int err = 0;
+	struct flash_pages_info page_info_start, page_info_end;
+	uint32_t erase_size;
+	int rc;
+
+	if (!len) {
+		return 0;
+	}
+
+	if (!qspi_flash_ra_valid(QSPI_NOR_FLASH_SIZE, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (len % QSPI_ERASE_BLK_SZ != 0) {
+		return -EINVAL;
+	}
+
+	rc = flash_get_page_info_by_offs(dev, offset, &page_info_start);
+	if ((rc != 0) || (offset != page_info_start.start_offset)) {
+		LOG_ERR("qspi: The offset 0x%lx is not aligned with the starting sector",
+			(long)offset);
+		return -EINVAL;
+	}
+
+	rc = flash_get_page_info_by_offs(dev, (offset + len), &page_info_end);
+	if ((rc != 0) || ((offset + len) != page_info_end.start_offset)) {
+		LOG_ERR("qspi: The size %u is not aligned with the ending sector", len);
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+	while (len > 0) {
+
+		if (len < BLOCK_SIZE_32K) {
+			erase_size = BLOCK_SIZE_4K;
+		} else if (len < BLOCK_SIZE_64K) {
+			erase_size = BLOCK_SIZE_32K;
+		} else {
+			erase_size = BLOCK_SIZE_64K;
+		}
+		err = R_QSPI_Erase(&qspi_data->qspi_ctrl,
+				   (uint8_t *)(QSPI_DEVICE_START_ADDRESS + offset), erase_size);
+		if (err) {
+			LOG_ERR("qspi: Erase failed");
+			err = -EIO;
+			break;
+		}
+
+		err = get_flash_status(dev);
+		if (err) {
+			LOG_ERR("qspi: failed to get status for QSPI operation");
+			err = -EIO;
+			break;
+		}
+
+		offset += erase_size;
+		len -= erase_size;
+	}
+	release_device(dev);
+
+	return err;
+}
+
+static int qspi_flash_ra_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	if (!qspi_flash_ra_valid(QSPI_NOR_FLASH_SIZE, offset, len)) {
+		return -EINVAL;
+	}
+
+	if (!len) {
```

Comment:
Here you fail operation of -EINVAL, even if len == 0 would cause nothing happening at all,
on the other side in `qspi_flash_ra_write` you first check len == 0 and bypass any more checks as there is no more work to.

Consider having one flow everywhere, for example checking the  len == 0 and just exit as there is no work to do and actually no possibility to failure